Alphabetical ID (redo #1178)

esoteric-ephemera opened the pull request:
**Mistakenly closed #1178 by rebasing git history. This PR expands upon it.**

Defines a new `AlphaID` class that could eventually replace / contain the current `MPID` class. 

From internal discussions, the benefit of the `MPID` system was brevity when the system was relatively new. "mp-149" is easy to remember whereas the current batch of MPIDs are > 3,000,000. 

To replace / augment the current `MPID` system, we need an identifier that:
- Can be sorted
- Can mint an $N+1$ ID given that $N$ IDs have currently been assigned
- Is easy to remember

From this, it was suggested that an alphabetical string (to avoid clashes with the current MPIDs, no numbers can be used) could be used instead. The integer value of this string would essentially be taken as base-26 representation, i.e.:
- "a" = $0 \times 26^0 = 0$
- "bc" = $1 \times 26^1 + 2 \times 26^0 = 28$
- "aaft" = $0 \times 26^3 + 0 \times 26^2 + 5 \times 26^1 + 19 \times 26^0 = 149$

The current implementation supports these features, as well as addition, subtraction, ==, <, > with other AlphaID, int, and the existing MPID. Includes convenience constructor from MPID to AlphaID.

**Tests included for these features.**

To make these easy to remember, we may want to set the pad length (the number of leading zeroes or "a" characters) to be at least 6, which would give us $26^6 = 308,915,776$ total task IDs (minus the ~3,100,000 that have currently been assigned).

Suggestions / discussion are welcome

tsmathis commented:
one more comment on here @esoteric-ephemera, adding `AlphaID` as a model field breaks pydantic [branch ci run](https://github.com/materialsproject/emmet/actions/runs/15120381760/job/42501168984):

```shell
pydantic.errors.PydanticSchemaGenerationError: Unable to generate pydantic-core schema for <class 'emmet.core.mpid.AlphaID'>. Set `arbitrary_types_allowed=True` in the model_config to ignore this error or implement `__get_pydantic_core_schema__` on your type to fully support it.

If you got this error by calling handler(<some type>) within `__get_pydantic_core_schema__` then you likely need to call `handler.generate_schema(<some type>)` since we do not call `__get_pydantic_core_schema__` on `<some type>` otherwise to avoid infinite recursion.
```

`MPID` has this dealt with the three methods here: https://github.com/materialsproject/emmet/blob/f1f38ad8765b53c3085dbca3d173571f535b94fa/emmet-core/emmet/core/mpid.py#L100

esoteric-ephemera commented:
@tsmathis fixed the pydantic serialization and added tests for this. Also added a `_calculation_to_props_dict` method to `Trajectory` to convert a `emmet.core.vasp.calculation.Calculation` to the format `Trajectory._from_dict` expects

esoteric-ephemera commented:
@tschaume  this should be good to go - it's a bigger change but mostly independent of any existing structures in emmet-core. Do you want to look it over before merging?

esoteric-ephemera commented on emmet-core/emmet/core/mpid.py:
```diff
@@ -74,7 +83,7 @@ def __str__(self):
+    def __lt__(self, other: MPID | int | str):
         other_parts = MPID(other).parts
 
         if self.parts[0] != "" and other_parts[0] != "":
```
There's also a bug in `MPID.__gt__` that I'll try to fix:

```
MPID('100') < MPID(100)
>>> False

MPID('100') > MPID(100)
>>> True

MPID('100') == MPID(100)
>>> True
```

esoteric-ephemera commented on emmet-core/emmet/core/mpid.py:
```diff
@@ -225,3 +231,311 @@ def validate(cls, __input_value: Any, _: core_schema.ValidationInfo):
+            ):
+                separator = list(non_alpha_char)[0]
+                prefix, identifier = identifier.split(separator)
+            elif len(non_alpha_char) > 1:
```
Good catch, will add an exception for this. Note that `AlphaID('')` is valid input and gives a zero-valued identifier

esoteric-ephemera commented on emmet-core/emmet/core/mpid.py:
```diff
@@ -225,3 +231,311 @@ def validate(cls, __input_value: Any, _: core_schema.ValidationInfo):
+        will not add the two. Only checks the separator if prefixes are both non-null.
+
+        Args:
+            other (str or int) : the value to add to the current identifier.
```
yeah just an outdated docstr

esoteric-ephemera commented on emmet-core/emmet/core/mpid.py:
```diff
@@ -225,3 +231,311 @@ def validate(cls, __input_value: Any, _: core_schema.ValidationInfo):
+        else:
+            test = other
+
+        if isinstance(test, AlphaID) and (
```
merged this into the staticmethod `_coerce_value` used across add, sub, lt, gt

esoteric-ephemera commented on emmet-core/emmet/core/trajectory.py:
```diff
@@ -194,54 +269,131 @@ def _from_dict(
+                if icr > 0:
+                    trajs.append(cls._from_dict(props, **old_meta, **kwargs))  # type: ignore[arg-type]
+
+                props = new_props.copy()
```
To be safe, probably yes, but I don't think there have been parsing issues so far using shallow copy

tschaume commented on emmet-core/emmet/core/oxidation_states.py:
```diff
@@ -36,50 +49,97 @@ class OxidationStateDoc(PropertyDoc):
-    def from_structure(cls, structure: Structure, material_id: MPID | None = None, **kwargs):  # type: ignore[override]
-        # TODO: add check for if it already has oxidation states, if so pass this along unchanged ("method": "manual")
+    def from_structure(
+        cls, structure: Structure, material_id: MPID | AlphaID, **kwargs
```
Why was `material_id` optional before (i.e. could be `None`)?

tsmathis replied:
dropping the optional on mpids would introduce break this https://github.com/materialsproject/emmet/issues/1205

tschaume commented on emmet-core/emmet/core/xrd.py:
```diff
@@ -113,5 +119,6 @@ def from_target(
             edge=edge,
             min_two_theta=min_two_theta,
             max_two_theta=max_two_theta,
+            symprec=symprec,
```
Great catch! I wonder what effect this bug has on our current XRD docs (if any).

tschaume commented on test_files/elasticity/SiC_tasks.json.gz:
It might make sense to outsource all (larger) binary files to `git lfs`.

tschaume reviewed (changes requested):
Looks ready! Only some previous unresolved comments left in `emmet-core/emmet/core/mpid.py` plus a few questions I added upon second review.

esoteric-ephemera commented on emmet-core/emmet/core/qchem/task.py:
```diff
@@ -120,6 +121,10 @@ class TaskDocument(BaseTaskDocument, MoleculeMetadata):
     Definition of a Q-Chem task document
     """
 
+    task_id: MPID | MPculeID | None = Field(
```
It inherits a `task_id` field from `BaseTaskDocument`, but the typing of `task_id` in `BaseTaskDocument` is flexible enough to accommodate MPID, AlphaID, or MPculeID. Just over-setting the field here in the molecules docs to narrow the scope to exclude AlphaID (for now at least)

esoteric-ephemera commented:
@tschaume and @kbuma added very minimal `pyproject.toml` files to the namespaces. Could you take a look and make sure they seem OK?

This will eventually be necessary, but I added them mostly to get around having to do `pip install -e ./emmet-X --config-settings editable_mode=compat`

Not sure we want or really need a full transition to pyproject-based name spaces, but the option is there
